black-jack-redo.rb

```diff
@@ -25,88 +27,212 @@ def print_with_stars(string)
+	else
+		user_hand = user_hand + "The dealers cards add up to #{hand_value} points!" + "\n"
+	end
+	return user_hand, hand_value
 end
 
-def dealer(count)
-	number = rand(4)
-	number = 21 - number
-	if count == number
-		return "You and the dealer both had #{number}, it's a tie!"
-	elsif count < number
-		return "The dealer had #{number} and you had #{count}, you lose!"
-	else count > number
-		return "The dealer had #{number} and you had #{count}, you win!"
+def get_value_of_hand(current_hand)
+	user_hand, hand_value = viewhand(current_hand, 'true')
+	return hand_value
+end
+
+def compare_player_dealer_points(player_points, dealer_points)
+	if player_points == dealer_points
+		winner = "You and the dealer both had #{dealer_points}, dealer wins!"
+	elsif player_points < dealer_points
+		winner = "The dealer had #{dealer_points} and you had #{player_points}, you lose!"
+	elsif player_points > dealer_points
+		winner = "The dealer had #{dealer_points} and you had #{player_points}, you win!"
+	else
+		puts "Something went wrong when determining who won!!"
 	end
+
+	return winner
 end
 
-def draw(array)
-	number = array.length - 1
-	puts array[number]
-	return array[number]
+def who_wins(current_deck, current_hand, dealer_hand)
+	user_hand, hand_value = viewhand(current_hand, 'true')
+	updated_dealer_hand, dealer_value = viewhand(dealer_hand, 'false')
+	if (dealer_value < 17) && (dealer_value < hand_value)
+		updated_dealer_hand, dealer_value = viewhand(dealer_hand, 'false')
+		print_with_stars(updated_dealer_hand)
+		while (dealer_value < 17) && (dealer_value < hand_value)
+			updated_dealer_hand, bool = drawone_card(current_deck, dealer_hand)
+			current_deck.pop
+			updated_dealer_hand, dealer_value = viewhand(dealer_hand, 'false')
+			print_with_stars(updated_dealer_hand)
+			if bool == true
+				who_wins_string = 'The dealer bust! You win!'
+				print_with_stars(who_wins_string)
+				return
+			end
+		end
+
+		who_wins_string = user_hand + updated_dealer_hand + compare_player_dealer_points(hand_value, dealer_value)
+	else
+		who_wins_string = user_hand + updated_dealer_hand + compare_player_dealer_points(hand_value, dealer_value)
+	end
+
+	print_with_stars(who_wins_string)
 end
 
-def drawone(array, cards, count)
-	bust = false
-	number = rand(52)
-	if array[number][:dealt] == true
-		draw(array, cards)
+def hand_contains_ace(current_hand)
+	if (current_hand.include?('SA')) || (current_hand.include?('HA')) || (current_hand.include?('DA') )|| (current_hand.include?('CA'))
+		return true
 	else
-		array[number][:dealt] = true
-		puts 'Your were dealt: '
-		puts array[number][:card] + ' of ' + array[number][:suit]
-		value = array[number][:value] + count
-		puts "Your cards add up to #{value} points!"
-		if value > 21
-			puts "Oh no! Your total is now #{value} which is greater than 21, you lose!"
-			bust = true
+		return false
+	end
+end
+
+def ace_handler(current_hand)
+	current_hand.length.times do |array_index|
```

Comment:
There are a few rare times when it's okay to use a single letter variable, because the RoR community has created conventions that are universally accepted. Replace `array_index` with `i`.